models/src/main/scala/coop/rchain/models/rholangn/Collection.scala

```diff
@@ -69,9 +77,13 @@ object ESetN {
   * @param remainder gives support to use ... in the set construction and deconstruction e.g. {"a":1, "b":2 ... rest}.
   *                  It's defined as optional variable.
   */
-final class EMapN(private val ps: TreeMap[ParN, ParN], val remainder: Option[VarN])
-    extends CollectionN {
-  def sortedPs: Seq[(ParN, ParN)] = ps.toSeq
+final class EMapN(val ps: TreeMap[ParN, ParN], val remainder: Option[VarN]) extends CollectionN {
+  // Sorted by objects hash which is memoized as Rho type
+  val psSorted: Eval[Seq[(ParN, ParN)]] =
+    ps.toSeq
+      .traverse(pp => pp.bimap(_.rhoHash, _.rhoHash).mapN(_ ++ _).map((pp, _)))
+      .map(_.sortBy(_._2).map(_._1))
+      .memoize
```

Comment:
Could you explain why this memoizing is beneficial here?

Reply:
Previously internal TreeMap has been sorted against hash on insertion, so no need to have sorted version

Reply:
I'll change collection to use TreeXXX structure which is already sorted.

Memoization is used to prevent sorting when it need to be read multiple times. It's cheap because elements are not copied in memory.